code/modules/jobs/job_types/security_officer.dm

```diff
@@ -149,11 +205,228 @@ GLOBAL_LIST_INIT(available_depts, list(SEC_DEPT_ENGINEERING, SEC_DEPT_MEDICAL, S
+	 * before the proc is ever called.
+	*/
+
+	preferences = preferences.Copy()
+	departments = departments.Copy()
+
+	var/distribution = get_distribution(preferences.len, departments.len)
+	var/selection[preferences.len]
+
+	var/list/grouped = list()
+	var/list/biggest_group
+	var/biggest_preference
+	var/list/indices = list()
+
+	for (var/index in 1 to preferences.len)
+		indices += index
+
+		var/preference = preferences[index]
+		if (!(preference in grouped))
+			grouped[preference] = list()
+		grouped[preference] += index
+
+		var/list/preferred_group = grouped[preference]
+
+		if (preference != SEC_DEPT_NONE && (isnull(biggest_group) || biggest_group.len < preferred_group.len))
+			biggest_group = grouped[preference]
+			biggest_preference = preference
+
+	if (isnull(biggest_group))
+		preferences[1] = pop(departments)
+		return get_officer_departments(preferences, departments)
+
+	if (biggest_group.len >= distribution[1])
+		for (var/index in 1 to distribution[1])
+			selection[biggest_group[index]] = biggest_preference
+
+		if (biggest_group.len > distribution[1])
+			for (var/leftover in (distribution[1] + 1) to biggest_group.len)
+				preferences[leftover] = SEC_DEPT_NONE
+	else
+		var/needed = distribution[1] - biggest_group.len
+		if (LAZYLEN(LAZYACCESS(grouped, SEC_DEPT_NONE)) >= needed)
+			for (var/candidate_index in biggest_group)
+				selection[candidate_index] = biggest_preference
+
+			for (var/index in 1 to needed)
+				selection[grouped[SEC_DEPT_NONE][index]] = biggest_preference
+		else
+			var/least_popular_index = grouped[grouped.len]
+			if (least_popular_index == SEC_DEPT_NONE)
+				least_popular_index = grouped[grouped.len - 1]
+			var/least_popular = grouped[least_popular_index]
+			for (var/candidate_index in least_popular)
+				preferences[candidate_index] = SEC_DEPT_NONE
+
+	// Remove all members of the most popular candidate from the list
+	for (var/chosen in 1 to selection.len)
+		if (selection[chosen] == biggest_preference)
+			indices -= chosen
+			preferences[chosen] = null
+
+	listclearnulls(preferences)
+
+	departments -= biggest_preference
+
+	if (grouped.len != 1)
+		var/list/next_step = get_officer_departments(preferences, departments)
+		for (var/index in 1 to indices.len)
+			var/place = indices[index]
+			selection[place] = next_step[index]
+
+	return selection
+
+/proc/get_distribution(candidates, departments)
+	var/number_of_twos = min(departments, round(candidates / 2))
+	var/redistribute = candidates - (2 * number_of_twos)
+
+	var/distribution[max(1, number_of_twos)]
+
+	for (var/index in 1 to number_of_twos)
+		distribution[index] = 2
+
+	for (var/index in 0 to redistribute - 1)
+		distribution[(index % departments) + 1] += 1
+
+	return distribution
+
+/proc/get_new_officer_distribution_from_late_join(
+	preference,
+	list/departments,
+	list/distribution,
+)
+	/**
+	 * For late joiners, we're forced to put them in an alone department at some point.
```

Comment:
Why not make this the autodoc? Same for the above comment I suppose

Reply:
These are implementation details, auto-docs are for consumers.